File: class093_TreeDynamicProgramming/Code03_DiameterOfBinaryTree.java
package class078;

// 二叉树的直径
// 给你一棵二叉树的根节点，返回该树的直径
// 二叉树的 直径 是指树中任意两个节点之间最长路径的长度
// 这条路径可能经过也可能不经过根节点 root
// 两节点之间路径的 长度 由它们之间边数表示
// 测试链接 : https://leetcode.cn/problems/diameter-of-binary-tree/
//
// 解题思路:
// 1. 使用树形动态规划（Tree DP）的方法
// 2. 对于每个节点，我们需要知道以下信息：
//    - 以该节点为根的子树的最大深度（高度）
//    - 以该节点为根的子树的直径
// 3. 递归处理左右子树，综合计算当前节点的信息
// 4. 对于每个节点，经过该节点的最长路径 = 左子树的最大深度 + 右子树的最大深度
//    整个树的直径 = max(左子树直径, 右子树直径, 经过当前节点的最长路径)
//
// 时间复杂度: O(n) - n为树中节点的数量，需要遍历所有节点
// 空间复杂度: O(h) - h为树的高度，递归调用栈的深度
// 是否为最优解: 是，这是计算二叉树直径的标准方法
public class Code03_DiameterOfBinaryTree {

	// 不要提交这个类
	public static class TreeNode {
		public int val;
		public TreeNode left;
		public TreeNode right;
	}

	// 提交如下的方法
	public static int diameterOfBinaryTree(TreeNode root) {
		return f(root).diameter;
	}

	public static class Info {
		// 以当前节点为根的子树的直径
		public int diameter;
		// 以当前节点为根的子树的最大深度（高度）
		public int height;

		public Info(int a, int b) {
			diameter = a;
			height = b;
		}

	}

	public static Info f(TreeNode x) {
		// 基本情况：空节点
		if (x == null) {
			// 空树的直径为0，高度为0
			return new Info(0, 0);
		}
		
		// 递归处理左右子树
		Info leftInfo = f(x.left);
		Info rightInfo = f(x.right);
		
		// 计算当前子树的信息
		// 当前子树的高度 = max(左子树高度, 右子树高度) + 1
		int height = Math.max(leftInfo.height, rightInfo.height) + 1;
		
		// 当前子树的直径 = max(左子树直径, 右子树直径, 经过当前节点的最长路径)
		// 经过当前节点的最长路径 = 左子树高度 + 右子树高度
		int diameter = Math.max(leftInfo.diameter, rightInfo.diameter);
		diameter = Math.max(diameter, leftInfo.height + rightInfo.height);
		
		// 返回当前节点的信息
		return new Info(diameter, height);
	}

}

File: class093_TreeDynamicProgramming/Code03_DiameterOfBinaryTree.cpp
// 二叉树的直径 (Diameter of Binary Tree)
// 题目描述:
// 给定一棵二叉树，你需要计算它的直径长度。
// 一棵二叉树的直径长度是任意两个结点路径长度中的最大值。
// 这条路径可能穿过也可能不穿过根结点。
// 测试链接 : https://leetcode.cn/problems/diameter-of-binary-tree/

// 二叉树节点定义
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    int maxDiameter; // 存储最大直径
    
public:
    // 主函数：计算二叉树的直径
    int diameterOfBinaryTree(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        
        maxDiameter = 0;
        maxDepth(root);
        return maxDiameter;
    }

private:
    // 计算树的最大深度，同时更新最大直径
    int maxDepth(TreeNode* node) {
        if (node == nullptr) {
            return 0;
        }
        
        // 计算左右子树的最大深度
        int leftDepth = maxDepth(node->left);
        int rightDepth = maxDepth(node->right);
        
        // 更新最大直径：经过当前节点的最长路径 = 左子树深度 + 右子树深度
        maxDiameter = custom_max(maxDiameter, leftDepth + rightDepth);
        
        // 返回当前节点的最大深度
        return custom_max(leftDepth, rightDepth) + 1;
    }
    
    // 自定义max函数，避免使用标准库
    int custom_max(int a, int b) {
        return (a > b) ? a : b;
    }
};